- Initial dashboard widget system

atsanna opened the pull request:
A little color in the dashboard :)

![widgets](https://user-images.githubusercontent.com/119703/144754056-43be46ec-aaf3-488e-b66d-740e48e64dac.png)

Following the way you created the menus, I tried to create a widget system that, via Dependency Injection, allows you to add statistics, graphs, etc. on the dashboard

In the Bonfire class I added:
```
	/**
	 * Creates any admin-required widgets so they're
	 * available to use by any modules.
	 */
	private function setupWidgets()
	{
		$widgets =  service('widgets');

		$widgets->createWidget(Stats::class, "stats");
		$widgets->widget("stats")
			->createCollection("users");
		$widgets->widget("stats")
			->createCollection("groups");
		$widgets->widget("stats")
			->createCollection("other");
	}
```

The configuration takes place within each single module, which can decide to expose a widget in the isAdmin () function

For example, in the user module I added:

```
		// Settings widgets stats on dashboard
		$widgets = service('widgets');
		$users = new UserModel();
		$statsItem = new StatsItem([
			'bgColor' => 'bg-blue',
			'title' => 'Users',
			'value' => $users->countAll(),
			'url' => ADMIN_AREA . '/users',
			'faIcon' => 'fa fa-user',
		]);
		$widgets->widget("stats")->collection('users')->addItem($statsItem);
```

It is an initial version, but working.

Graphics are adaptive on a per-device basis.

What do you think?

atsanna committed:
- Initial dashboard widget system

atsanna committed:
- remove unused namespace
- remove spaces

atsanna committed:
- add some default values

atsanna committed:
- add type

atsanna committed:
- fix type

atsanna committed:
- changed the user icon to widget-stats

atsanna committed:
- changed the title of the user groups to widget-stats

atsanna committed:
- add partial testing

lonnieezell commented on bonfire/Views/Widgets/_stats.php:
```diff
@@ -0,0 +1,19 @@
+<?= asset_link( 'admin/css/widgets.css', 'css'); ?>
```
Would prefer the CSS is part of the theme. So adding it to the dashboard page, under the `styles` section would be more appropriate, I believe.

lonnieezell commented on bonfire/Bonfire.php:
```diff
@@ -84,6 +86,23 @@ private function setupMenus()
+
+		$widgets->createWidget(Stats::class, "stats");
+		$widgets->widget("stats")
+			->createCollection("users");
```
Creating groups for users, etc in the main Bonfire file doesn't seem appropriate to me. Since those are specific to a module I think just having a `stats` collection is fine here. That leaves the door open for other widget types that would have their own collections. Then user users and groups modules would just need to add their widgets to the stats collection.

atsanna replied:
You're absolutely right: I forgot to specify that I've added a bunch of useless collections just to show that it works similar to your menu management

lonnieezell reviewed (changes requested):
Really nice start. Thanks! A couple of comments: 

1. What is your thinking on the future of the widgets? Mainly asking because I'm a little confused by the need for both a Widgets Manager and the stats collection. If you plan on providing other widget types then I think it makes sense. 
2. There's a few code style things, but if you run `composer style` in the web root it should run php-cs-fixer and update the code so the styling is consistent.

I really like the addition (and choice) of colors you've provided, and the use of CSS variables for it all.

atsanna commented:
> Really nice start. Thanks! A couple of comments:
> 
> 1. What is your thinking on the future of the widgets? Mainly asking because I'm a little confused by the need for both a Widgets Manager and the stats collection. If you plan on providing other widget types then I think it makes sense.
> 2. There's a few code style things, but if you run `composer style` in the web root it should run php-cs-fixer and update the code so the styling is consistent.
> 
> I really like the addition (and choice) of colors you've provided, and the use of CSS variables for it all.

Glad you liked the draft.

As for your first comment, I imagined the dashboard as a page that, in addition to collecting and displaying statistical data, gives the possibility to aggregate data from multiple tables and / or from multiple modules.

From my humble point of view, having only simple predefined counters or graphs with data from a single table makes little sense in the dashboard.

The dashboard should also allow for data analysis based on useful criteria chosen by the user.

If we consider a shop, instead of displaying only the number of products sold each month, perhaps it is more useful to have the possibility to understand which are the most sold or requested product categories based on the location.

If we consider a real estate agency, perhaps I would like to understand which types of houses are most in demand based on the clients' budget.

In a management of the building stock, which are the types of maintenance that have a greater weight on costs.

A crm will have to analyze leads, opportunities, sales, etc.

Some applications may need to display some widgets on the dashboard, even within the datasheet of a single product, a single building, etc.

Each application needs specific tools and, of course, no one can know them all.
But if we have a flexible enough solution, anyone could add what they need.

The first widget I created as an example, just to give you an idea of ​​how it should work, is just a counter with a shortcut that allows me to open the list of elements of a module.

The next step could be to allow for charting.
On this point it would be necessary to choose together which instrument to use.
Here are some examples:
https://www.chartjs.org/
https://www.flotcharts.org/
https://d3js.org/
https://apexcharts.com/
...

After charts, you might think about displaying tables with aggregated data from multiple tables or modules.

Other examples of widget types could be:
- last 10 notifications
- last 10 critical logs
- featured posts or comments
- To-do list
- Calendar
- geographic map of visitors
...

Subsequently it could be expanded giving the possibility to each individual user to customize the dashboard according to their needs.

This idea, very ambitious for me since I have never developed anything like it, is just an idea; a hypothesis of how it could evolve, obviously one small step at a time ...

It is clear that you have a plan to develop Bonfire and I do not know your dashboard idea.

If you tell me your idea, within the limits of my knowledge and ability, I might try to write exactly what you need, the way you need it.

As for the second comment referring to the code, I will try to do what is necessary. I've never used php-cs-fixer, but I'll be happy to learn a new tool ...

atsanna committed:
- moved css link to appropriate location
- reconfigured user and group widget position
- removed unnecessary collections in Bonfire main file

atsanna commented:
> 2. There's a few code style things, but if you run `composer style` in the web root it should run php-cs-fixer and update the code so the styling is consistent.

I tried to run the composer style command from the same location where I normally run the composer update command.

The composer style command returns this error:

```
composer style
> php-cs-fixer fix --verbose --ansi --using-cache=no
PHP CS Fixer 3.3.2 Trinacria by Fabien Potencier and Dariusz Ruminski
Runtime: PHP 8.0.8
Loaded config default.

In Finder.php line 614:
                                                                                 
  [LogicException]                                                               
  You must call one of in() or append() methods before iterating over a Finder.  

```
Could you tell me how to solve the problem?

In the meantime I'm trying to find the solution on Google ...

[ **Update** ]
I think I understand that it cannot find the _.php-cs-fixer.php_ file which should be in the web root

[**Update 2**]
I had guessed the problem:

I downloaded the file from this link https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/.php-cs-fixer.dist.php and everything worked.

The problem is that it adds a header that isn't correct:
```
/ *
    * This file is part of PHP CS Fixer.
    *
    * (c) Fabien Potencier <fabien@symfony.com>
    * Dariusz Rumiński <dariusz.ruminski@gmail.com>
    *
    * This source file is subject to the bundled MIT license
    * with this source code in the LICENSE file.
    * /
```

But I guess you have your own .php-cs-fixer.php file for use in this project - could you share it?

atsanna committed:
- add collect function
- add some test

atsanna committed:
- add some test

atsanna committed:
- add icon animation

atsanna committed:
- improved icon animation

atsanna committed:
- add the Guides for widgets and statistics

atsanna committed:
- fix the typing in the statistics guide

lonnieezell commented:
Actually - I'm getting the same error. Guess that never quite got setup right. I'll try to get that working tonight for you. 

And I agree with long term plans for the dashboard. Sounds great. I hadn't gotten as far as planning that yet. :) I've used flot a lot in the past and it was fine. I'm amazed at how much can be done with just CSS these days and wonder if we could use CSS for the handful of most basic charts and not require another JS lib? Something to think on, anyway.

atsanna commented:
I would like the idea of being able to generate graphs only with css.
I don't know if you know this project: https://chartscss.org/
It doesn't have all the types of charts available right now, but as a starting point it doesn't seem too bad ...

atsanna committed:
Merge branch 'lonnieezell:develop' into dashboard_widgets

lonnieezell commented:
I hadn't heard of that lib before. Unfortunately, I don't think they're going to hit their targets of the pie/radial/etc before year end. Doesn't seem to be any activity in the last 4 months. 

I did a quick search, and Google Charts might be a good solution. It's HTML5/SVG and backed by Google. Looks pretty complete at a quick glance. https://developers.google.com/chart/   Thoughts?

atsanna commented:
Certainly there is no shortage of graphic types :)
I've never used it, but it's fine by me.

lonnieezell commented:
Having looked a little bit more, I say lets go with ChartJS. Google was JS anyway, and I like the look of these charts better, and it seems to have a good ecosystem.

lonnieezell commented:
I've added the .php-cs-fixer file and did the initial run. If you could run that on your code we can get it merged. Thanks

lonnieezell committed:
Revert styling

lonnieezell committed:
Redo styling with correct header

lonnieezell committed:
Consent settings page.

atsanna committed:
- Added php-cs-fixer file and initial cleanup.

atsanna committed:
Revert "- fix the typing in the statistics guide"

This reverts commit 67d03b1dbe2dccbf521e9b449e76a68fa5d477bf.

atsanna committed:
Merge remote-tracking branch 'origin/dashboard_widgets' into dashboard_widgets

atsanna committed:
Merge branch 'dashboard_widgets' into lonnie-develop

# Conflicts:
#	bonfire/Bonfire.php
#	bonfire/Config/BaseModule.php
#	bonfire/Filters/OnlineCheck.php
#	bonfire/Modules/Assets/Helpers/assets_helper.php
#	bonfire/Modules/Consent/Filters/ConsentFilter.php
#	bonfire/Modules/Users/Module.php
#	bonfire/Modules/Users/Views/user_settings.php

atsanna committed:
Merge remote-tracking branch 'origin/lonnie-develop' into lonnie-develop

atsanna committed:
- Added

atsanna committed:
- styling

atsanna committed:
- missing service

atsanna committed:
- Apply styling with composer style

atsanna commented:
I hope I did the right things - I'm not very familiar with git commands yet...

atsanna commented:
> Having looked a little bit more, I say lets go with ChartJS. Google was JS anyway, and I like the look of these charts better, and it seems to have a good ecosystem.

Great: I start playing with ChartJS :)

atsanna committed:
Merge branch 'lonnieezell:develop' into dashboard_widgets

atsanna commented:
![Charts](https://user-images.githubusercontent.com/119703/145616951-3d3449fa-d999-45e7-aa9a-6573318fe533.png)

Hi Lonnie, are you all right?

As you can see from the image, you start to see the graphs :)

Their configuration always takes place within the Module.php file

This is a configuration example:
```php
		$groupsData = db_connect()->table('auth_groups_users')
			->select("group")
			->selectCount('user_id')
			->groupBy("group")
			->get()
			->getResultArray();

		//Prepare Data
		$data = [];
		$label = [];
		foreach ($groupsData as $el){
			$data[] = $el['user_id'];
			$label[] = $el['group'];
		}

		$chartItem = new ChartsItem([
			'title'   => 'Groups',
			'type'   => 'line',
			'cssClass'   => 'col-6',
			'data' => $data,
			'label' =>$label,
		]);
		$widgets->widget('charts')->collection('charts')->addItem($chartItem);
```

Another way to assign the dataset is as follows:
```php
		$statsItem2 = new ChartsItem([
			'title'   => 'User2',
			'type'   => 'bar',
			'cssClass'   => 'col-6'
		]);
		$statsItem2->addDataset('auth_groups_users', 'group', 'user_id');
		$widgets->widget('charts')->collection('charts')->addItem($statsItem2);
```

For now this is a first version that accepts a single dataset and only supports 5 chart types: line, bar, donut, pie and poolarArea.

I have yet to write the tests and documentation ...

Once completed, would you prefer me to add to this PR or create a new one?

lonnieezell commented:
That's looking great. Adding the tests and docs to this PR is good. Thanks!

atsanna committed:
- Add initial graphic widgets

atsanna committed:
- Add some initial PHPDoc

atsanna committed:
Merge branch 'lonnieezell:develop' into dashboard_widgets

atsanna committed:
- Fix sidebar width problem with chartjs script

atsanna committed:
Merge remote-tracking branch 'origin/dashboard_widgets' into dashboard_widgets

atsanna committed:
- Update charts guide

atsanna committed:
- Add some tests on the charts

atsanna committed:
- Fix code style

atsanna committed:
- Fix Style

atsanna commented:
![Settings](https://user-images.githubusercontent.com/119703/145731355-5b47772f-8b19-4936-9120-cf5eaa9491b7.png)

The charts seem to work. I also added a panel for general chart settings.

I think you can put this part together - what do you think?

atsanna committed:
- Fix Style

atsanna committed:
- Fix Style and missing PHPDoc

atsanna committed:
- Typo correction

atsanna committed:
- Added Show Legend setting

atsanna committed:
- Added Missing break

atsanna committed:
- Update Documentation

atsanna committed:
- Added legend position setting
- Update Documentation

atsanna committed:
- Apply styling with composer style

atsanna committed:
- Removed console.log

atsanna committed:
- Add missed config files

atsanna committed:
- Update Charts Guide

atsanna committed:
- Added the ability to set a default color scheme to fill the chart

atsanna committed:
- Correction of text labels

lonnieezell commented:
@atsanna Just let me know when you're happy with this PR. I've been letting ride as long as updates keep getting pushed. :)

atsanna committed:
- New widget admin panel
- Updated the guides

atsanna commented:
![Settings_new](https://user-images.githubusercontent.com/119703/146052179-b7d687a3-8df6-4155-a711-e6363f9445f0.png)
I've revisited the widget admin panel - it looks neater to me now.

If you agree, I think this part can be joined. 

I hope you like the result.

atsanna committed:
- Typo correction

atsanna committed:
- schema color preview with htmx

atsanna committed:
- Added the ability to enable/disable individual widgets from the dashboard

atsanna committed:
- Added chart type

atsanna committed:
- Update Guide

atsanna committed:
- type correction

atsanna committed:
- Fix sidebar width problem with chartjs script

atsanna committed:
- Some minor corrections

atsanna committed:
- Minor changes to button labels

lonnieezell commented:
Merged! Thanks so much for this.
